src/core/net/dnssd_server.cpp

```diff
@@ -1287,6 +1386,574 @@ void Server::ResetUpstreamQueryTransaction(UpstreamQueryTransaction &aTxn, Error
+        info.mAction  = nextAction;
+        response.mMessage->Write(0, response.mHeader);
+
+        if (response.mMessage->Append(info) != kErrorNone)
+        {
+            response.SetResponseCode(Header::kResponseServerFailure);
+            response.Send(info.mMessageInfo);
+            continue;
+        }
+
+        // Take back ownership of `response.mMessage` as we still
+        // treat it as a `ProxyQuery`.
+
+        response.mMessage.Release();
+
+        // We place the `query` in a separate list and add it back to
+        // the main `mProxyQueries` list after we are done with the
+        // current iteration. This ensures that other entries in the
+        // `mProxyQueries` list are not updated or removed due to the
+        // DNS-SD platform callback being invoked immediately when we
+        // potentially start a browser or resolver to perform the
+        // `nextAction` for `query`.
+
+        nextActionQueries.Enqueue(query);
+    }
+
+    for (ProxyQuery &query : nextActionQueries)
+    {
+        nextActionQueries.Dequeue(query);
+
+        info.ReadFrom(query);
+
+        nextAction = info.mAction;
+
+        info.mAction = kNoAction;
+        info.UpdateIn(query);
+
+        Get<Server>().mProxyQueries.Enqueue(query);
+        Perform(nextAction, query, info);
+    }
+}
+
+Error Server::Response::AppendPtrRecord(const ProxyResult &aResult)
+{
+    mSection = kAnswerSection;
+
+    return AppendPtrRecord(aResult.mServiceInstance->mServiceInstance, aResult.mServiceInstance->mTtl);
+}
+
+Error Server::Response::AppendService(const ProxyResult &aResult)
+{
+    static const Section kSections[] = {kAnswerSection, kAdditionalDataSection};
+
+    Error   error      = kErrorNone;
+    Section srvSection = ((mType == kSrvQuery) || (mType == kSrvTxtQuery)) ? kAnswerSection : kAdditionalDataSection;
+    Section txtSection = ((mType == kTxtQuery) || (mType == kSrvTxtQuery)) ? kAnswerSection : kAdditionalDataSection;
+
+    for (Section section : kSections)
+    {
+        mSection = section;
+
+        if (mSection == kAdditionalDataSection)
+        {
+            VerifyOrExit(!(Get<Server>().mTestMode & kTestModeEmptyAdditionalSection));
+        }
+
+        if (srvSection == mSection)
+        {
+            SuccessOrExit(error = AppendSrvRecord(*aResult.mService));
+        }
+
+        if (txtSection == mSection)
+        {
+            SuccessOrExit(error = AppendTxtRecord(*aResult.mService));
+        }
+    }
+
+exit:
+    return error;
+}
+
+Error Server::Response::AppendSrvRecord(const Dnssd::Service &aService)
+{
+    DnsName fullHostName;
+
+    ConstructFullName(aService.mHostName, fullHostName);
+
+    return AppendSrvRecord(fullHostName, aService.mTtl, aService.mPriority, aService.mWeight, aService.mPort);
+}
+
+Error Server::Response::AppendTxtRecord(const Dnssd::Service &aService)
+{
+    return AppendTxtRecord(aService.mTxtData, aService.mTxtDataLength, aService.mTtl);
```

Comment:
We can have different TTLs for the two records. 
This may require a change to `Dnssd::Service` type to provide separate TTLs for this.
Can change this in a future push.

Reply:
Thanks for you answer! Looking forward for the changes :)